### DISCUSSION TITLE
Make it possible to use native brightness increase for lights

### DISCUSSION THREAD

**Discussion by recht** [home-assistant/architecture#603] [Entity Models]
Some lights natively support increasing or decreasing brightness instead of setting the absolute number. Currently, HA translates `brightness_step` to an absolute number instead of passing it down to supported implementations. The main downside of this is that it's not possible to stop transitions, as that's typically done by setting the step to `0` (this is the case for Hue and IKEA).

My suggestion is to add a feature flag on lights (`SUPPORTS_BRIGHTNESS_INC`), and if an implementation contains that feature then just pass `brightness_step` to the implementation and let it translate it to the native API.

My suggestion is already implemented in https://github.com/home-assistant/core/pull/54341 for Hue lights.

**Comment by balloob** [comment]
I wonder if we need this, or that we can also pass the `brightess_step_pct` to the turn_on command and have the turn_on command deal with it instead. No need for a new service call ?

**Reply by recht** [reply]
`step_pct` can be converted to `step`, so that's definitely also an option.

My proposal doesn't contain a new service call if that's what you mean - it's just passing more args down to the turn_on service implementations. The alternative is to add a new service call, like `light/stop_transition`, which arguably is more user friendly, and might even be more future proof, in case some systems support transitions with explicit cancel operations instead of overloading `bri_inc`.

**Reply by balloob** [reply]
Why would we require a new support flag if we can just pass it down anyway?

**Reply by recht** [reply]
For Hue lights it can be passed down, but I don't know if there are lights out there that do not support stepping - I assume there is since the original implementation did stepping in code.

**Reply by recht** [reply]
Any input here? I'm pretty sure a number of implementations do not support native transition, so I don't see how to avoid introducing some support flag.

**Reply by chishm** [reply]
I'm also interested in this, for the deconz component.

I think what balloob was saying is that if both the `brightness` and `brightness_step`/`brightness_step_pct` kwargs are passed to the `turn_on` function of the integration, then that function can use the `brightness_step` kwarg if it knows about it and can support it, or default to using the `brightness` kwarg as is currently done. Provided all `turn_on` functions accept a `**kwargs` argument this won't introduce any issues.

**Reply by recht** [reply]
And my point is is that right now, step is implemented in HA, so if we just pass down the args then stepping won't work for implementations that so not support it natively. Check out the linked PR to see the change I'm proposing, that should make it clear (unless I'm missing something)

**Reply by chishm** [reply]
I've checked out that PR and I think I understand what you're saying. However, `async_handle_light_on_service` can still do the brightness step calculation *and* pass the original step into the `turn_on` method, like so (notice `params.pop` has been replaced with `params[]`):
```py
async def async_handle_light_on_service(light, call):
    params = dict(call.data["params"])

    if params and (
        ATTR_BRIGHTNESS_STEP in params or ATTR_BRIGHTNESS_STEP_PCT in params
    ):
        brightness = light.brightness if light.is_on else 0

        if ATTR_BRIGHTNESS_STEP in params:
            brightness += params[ATTR_BRIGHTNESS_STEP]
        else:
            brightness += round(params[ATTR_BRIGHTNESS_STEP_PCT] / 100 * 255)

        params[ATTR_BRIGHTNESS] = max(0, min(255, brightness))
```

`light.__init__.py` will also need modification in `preprocess_turn_on_alternatives`, but it can be done.

Then the integration method can look like:

```py
class IntegrationLight(LightEntity):
    async def async_turn_on(self, **kwargs):
        if ATTR_BRIGHTNESS_STEP in kwargs:
            do_brightness_step()
        elif ATTR_BRIGHTNESS_STEP_PCT in kwargs:
            do_brightness_step_pct()
        elif ATTR_BRIGHTNESS in kwargs:
            do_brightness_set()
    
```
I do see why you've added `SUPPORT_BRIGHTNESS_INC`, since every other parameter of `async_handle_light_on_service` is filtered based on `supported_features` in `filter_turn_on_params`. TBH I think that's the more consistent approach, but probably not necessary.

**Reply by recht** [reply]
You're right, that could work, but requires correct ordering of the checks in the implementations. I guess right now that's not an issue since none of them check for step. I can try to do another PR with that approach and see if that might then get accepted.

**Reply by chishm** [reply]
Indeed, it is more fiddly and special-cased (compared to how all other light features are handled) to do it with multiple kwargs as I've illustrated above.

Perhaps @balloob can weigh in, so there's an official decision before trying for another PR.

**Reply by recht** [reply]
@balloob would you be ok with passing down the step params? I assume so since you suggested it (and I originally misunderstood the suggestion), but any chance you could confirm and close this thread, then I'd be happy to do a new PR?

**Reply by recht** [reply]
@balloob friendly ping?

**Reply by uberjay** [reply]
It'd be great to have some closure on this, as it's a real bummer. Because of this, I use hubitat for any button controller automation that needs a smooth brightness adjustment.

**Reply by elupus** [reply]
I think this make sense. Both pass a absolute value to the integration and the step value. If it can handle step, it should do so.
